File: app/backend/routers/watchlist.py
"""
/api/watchlist - ウォッチリストCRUD
holdings.pyパターンに準拠。
"""
import re
import logging
from fastapi import APIRouter, HTTPException, Depends, Query
from pydantic import BaseModel, field_validator
from typing import Optional, List
import main
from auth import require_auth

logger = logging.getLogger(__name__)
router = APIRouter()

_TICKER_RE = re.compile(r"^[A-Z0-9.\-]{1,10}$")


class WatchlistRecord(BaseModel):
    id: Optional[str] = None
    user_id: Optional[str] = None
    name: str = "メイン"
    tickers: List[str] = []
    is_default: bool = False
    created_at: Optional[str] = None
    updated_at: Optional[str] = None


class WatchlistCreate(BaseModel):
    name: str = "メイン"
    tickers: List[str] = []
    is_default: bool = False

    @field_validator("tickers")
    @classmethod
    def validate_tickers(cls, v: list[str]) -> list[str]:
        if len(v) > 50:
            raise ValueError("Maximum 50 tickers")
        return [t.upper() for t in v if _TICKER_RE.match(t.upper())]


class WatchlistUpdate(BaseModel):
    name: Optional[str] = None
    tickers: Optional[List[str]] = None
    is_default: Optional[bool] = None


class WatchlistsResponse(BaseModel):
    watchlists: List[WatchlistRecord]
    total: int


@router.get("", response_model=WatchlistsResponse)
async def get_watchlists(user_email: str = Depends(require_auth)):
    supabase = main.get_supabase()
    if not supabase:
        raise HTTPException(status_code=503, detail="Database not connected")
    try:
        result = (
            supabase.table("user_watchlists")
            .select("*")
            .eq("user_id", user_email)
            .order("is_default", desc=True)
            .order("name")
            .execute()
        )
        watchlists = [WatchlistRecord(**row) for row in result.data]
        return WatchlistsResponse(watchlists=watchlists, total=len(watchlists))
    except Exception as e:
        logger.exception("Watchlist get error")
        raise HTTPException(status_code=500, detail=str(e))


@router.post("", response_model=WatchlistRecord)
async def create_watchlist(body: WatchlistCreate, user_email: str = Depends(require_auth)):
    supabase = main.get_supabase()
    if not supabase:
        raise HTTPException(status_code=503, detail="Database not connected")
    try:
        data = {
            "user_id": user_email,
            "name": body.name,
            "tickers": body.tickers,
            "is_default": body.is_default,
        }
        result = supabase.table("user_watchlists").insert(data).execute()
        if not result.data:
            raise HTTPException(status_code=500, detail="Failed to create")
        return WatchlistRecord(**result.data[0])
    except HTTPException:
        raise
    except Exception as e:
        logger.exception("Watchlist create error")
        raise HTTPException(status_code=500, detail=str(e))


@router.put("/{watchlist_id}", response_model=WatchlistRecord)
async def update_watchlist(
    watchlist_id: str,
    body: WatchlistUpdate,
    user_email: str = Depends(require_auth),
):
    supabase = main.get_supabase()
    if not supabase:
        raise HTTPException(status_code=503, detail="Database not connected")
    try:
        update_data = {}
        if body.name is not None:
            update_data["name"] = body.name
        if body.tickers is not None:
            update_data["tickers"] = [t.upper() for t in body.tickers if _TICKER_RE.match(t.upper())]
        if body.is_default is not None:
            update_data["is_default"] = body.is_default
        if not update_data:
            raise HTTPException(status_code=400, detail="No update data")
        result = (
            supabase.table("user_watchlists")
            .update(update_data)
            .eq("id", watchlist_id)
            .eq("user_id", user_email)
            .execute()
        )
        if not result.data:
            raise HTTPException(status_code=404, detail="Watchlist not found")
        return WatchlistRecord(**result.data[0])
    except HTTPException:
        raise
    except Exception as e:
        logger.exception("Watchlist update error")
        raise HTTPException(status_code=500, detail=str(e))


@router.delete("/{watchlist_id}")
async def delete_watchlist(watchlist_id: str, user_email: str = Depends(require_auth)):
    supabase = main.get_supabase()
    if not supabase:
        raise HTTPException(status_code=503, detail="Database not connected")
    try:
        result = (
            supabase.table("user_watchlists")
            .delete()
            .eq("id", watchlist_id)
            .eq("user_id", user_email)
            .execute()
        )
        if not result.data:
            raise HTTPException(status_code=404, detail="Watchlist not found")
        return {"status": "deleted", "id": watchlist_id}
    except HTTPException:
        raise
    except Exception as e:
        logger.exception("Watchlist delete error")
        raise HTTPException(status_code=500, detail=str(e))


@router.post("/add-ticker")
async def add_ticker(
    ticker: str = Query(...),
    user_email: str = Depends(require_auth),
):
    """デフォルトウォッチリストにティッカー追加（auto-create）"""
    ticker = ticker.upper()
    if not _TICKER_RE.match(ticker):
        raise HTTPException(status_code=400, detail="Invalid ticker")
    supabase = main.get_supabase()
    if not supabase:
        raise HTTPException(status_code=503, detail="Database not connected")
    try:
        result = (
            supabase.table("user_watchlists")
            .select("*")
            .eq("user_id", user_email)
            .eq("is_default", True)
            .limit(1)
            .execute()
        )
        if result.data:
            wl = result.data[0]
            tickers = wl["tickers"] or []
            if ticker not in tickers:
                tickers.append(ticker)
                supabase.table("user_watchlists").update({"tickers": tickers}).eq("id", wl["id"]).eq("user_id", user_email).execute()
            return {"tickers": tickers}
        else:
            supabase.table("user_watchlists").insert({
                "user_id": user_email,
                "name": "メイン",
                "tickers": [ticker],
                "is_default": True,
            }).execute()
            return {"tickers": [ticker]}
    except HTTPException:
        raise
    except Exception as e:
        logger.exception("Add ticker error")
        raise HTTPException(status_code=500, detail=str(e))


@router.post("/remove-ticker")
async def remove_ticker(
    ticker: str = Query(...),
    user_email: str = Depends(require_auth),
):
    """デフォルトウォッチリストからティッカー削除"""
    ticker = ticker.upper()
    supabase = main.get_supabase()
    if not supabase:
        raise HTTPException(status_code=503, detail="Database not connected")
    try:
        result = (
            supabase.table("user_watchlists")
            .select("*")
            .eq("user_id", user_email)
            .eq("is_default", True)
            .limit(1)
            .execute()
        )
        if result.data:
            wl = result.data[0]
            tickers = [t for t in (wl["tickers"] or []) if t != ticker]
            supabase.table("user_watchlists").update({"tickers": tickers}).eq("id", wl["id"]).eq("user_id", user_email).execute()
            return {"tickers": tickers}
        return {"tickers": []}
    except Exception as e:
        logger.exception("Remove ticker error")
        raise HTTPException(status_code=500, detail=str(e))

File: app/worker/src/routes/watchlist.ts
/**
 * /api/watchlist — ウォッチリスト CRUD
 * ポート元: app/backend/routers/watchlist.py (227行)
 */

import type { Env } from '../env';
import { requireAuth, AuthError } from '../middleware/auth';
import { getSupabase } from '../lib/supabase';
import { jsonResponse, errorResponse } from '../lib/response';
import { parseJsonBody, isValidTicker } from '../lib/validation';

const TICKER_RE = /^[A-Z0-9.\-]{1,10}$/;

export async function handleWatchlist(
  request: Request,
  env: Env,
  cors: Record<string, string> | null,
): Promise<Response> {
  const supabase = getSupabase(env);
  const url = new URL(request.url);
  const path = url.pathname.replace('/api/watchlist', '') || '/';
  const method = request.method;

  try {
    const userId = await requireAuth(request, env, supabase);

    if (path === '/' && method === 'GET') return listWatchlists(supabase, userId, cors);
    if (path === '/' && method === 'POST') return createWatchlist(request, supabase, userId, cors);
    if (path === '/add-ticker' && method === 'POST') return addTicker(url, supabase, userId, cors);
    if (path === '/remove-ticker' && method === 'POST') return removeTicker(url, supabase, userId, cors);

    const idMatch = path.match(/^\/([^/]+)$/);
    if (idMatch) {
      if (method === 'PUT') return updateWatchlist(request, supabase, userId, idMatch[1], cors);
      if (method === 'DELETE') return deleteWatchlist(supabase, userId, idMatch[1], cors);
    }

    return errorResponse('Not Found', 404, cors);
  } catch (err) {
    if (err instanceof AuthError) return errorResponse(err.message, err.status, cors);
    return errorResponse('Internal server error', 500, cors);
  }
}

// GET /api/watchlist
async function listWatchlists(supabase: ReturnType<typeof getSupabase>, userId: string, cors: Record<string, string> | null) {
  const { data } = await supabase
    .table('user_watchlists').select('*').eq('user_id', userId)
    .order('is_default', { ascending: false }).order('name');
  const watchlists = data || [];
  return jsonResponse({ watchlists, total: watchlists.length }, 200, cors);
}

// POST /api/watchlist
async function createWatchlist(request: Request, supabase: ReturnType<typeof getSupabase>, userId: string, cors: Record<string, string> | null) {
  const body = await parseJsonBody(request);
  const name = typeof body.name === 'string' ? body.name : 'メイン';
  let tickers: string[] = [];
  if (Array.isArray(body.tickers)) {
    tickers = (body.tickers as string[])
      .map(t => String(t).toUpperCase())
      .filter(t => TICKER_RE.test(t))
      .slice(0, 50);
  }

  const { data, error } = await supabase.table('user_watchlists').insert({
    user_id: userId,
    name,
    tickers,
    is_default: body.is_default === true,
  }).select();

  if (error || !data?.length) return errorResponse('Failed to create', 500, cors);
  return jsonResponse(data[0], 201, cors);
}

// PUT /api/watchlist/:id
async function updateWatchlist(request: Request, supabase: ReturnType<typeof getSupabase>, userId: string, watchlistId: string, cors: Record<string, string> | null) {
  const body = await parseJsonBody(request);
  const updates: Record<string, unknown> = {};
  if (body.name !== undefined) updates.name = body.name;
  if (body.tickers !== undefined && Array.isArray(body.tickers)) {
    updates.tickers = (body.tickers as string[])
      .map(t => String(t).toUpperCase())
      .filter(t => TICKER_RE.test(t));
  }
  if (body.is_default !== undefined) updates.is_default = body.is_default;
  if (Object.keys(updates).length === 0) return errorResponse('No update data', 400, cors);

  const { data } = await supabase.table('user_watchlists')
    .update(updates).eq('id', watchlistId).eq('user_id', userId).select();
  if (!data?.length) return errorResponse('Watchlist not found', 404, cors);
  return jsonResponse(data[0], 200, cors);
}

// DELETE /api/watchlist/:id
async function deleteWatchlist(supabase: ReturnType<typeof getSupabase>, userId: string, watchlistId: string, cors: Record<string, string> | null) {
  const { data } = await supabase.table('user_watchlists')
    .delete().eq('id', watchlistId).eq('user_id', userId).select();
  if (!data?.length) return errorResponse('Watchlist not found', 404, cors);
  return jsonResponse({ status: 'deleted', id: watchlistId }, 200, cors);
}

// POST /api/watchlist/add-ticker
async function addTicker(url: URL, supabase: ReturnType<typeof getSupabase>, userId: string, cors: Record<string, string> | null) {
  const ticker = (url.searchParams.get('ticker') || '').toUpperCase();
  if (!isValidTicker(ticker)) return errorResponse('Invalid ticker', 400, cors);

  const { data: existing } = await supabase
    .table('user_watchlists').select('*').eq('user_id', userId).eq('is_default', true).limit(1);

  if (existing?.length) {
    const wl = existing[0];
    const tickers: string[] = wl.tickers || [];
    if (!tickers.includes(ticker)) {
      tickers.push(ticker);
      await supabase.table('user_watchlists').update({ tickers }).eq('id', wl.id).eq('user_id', userId);
    }
    return jsonResponse({ tickers }, 200, cors);
  }

  // デフォルトウォッチリスト自動作成
  await supabase.table('user_watchlists').insert({
    user_id: userId, name: 'メイン', tickers: [ticker], is_default: true,
  });
  return jsonResponse({ tickers: [ticker] }, 200, cors);
}

// POST /api/watchlist/remove-ticker
async function removeTicker(url: URL, supabase: ReturnType<typeof getSupabase>, userId: string, cors: Record<string, string> | null) {
  const ticker = (url.searchParams.get('ticker') || '').toUpperCase();

  const { data: existing } = await supabase
    .table('user_watchlists').select('*').eq('user_id', userId).eq('is_default', true).limit(1);

  if (existing?.length) {
    const wl = existing[0];
    const tickers = (wl.tickers || []).filter((t: string) => t !== ticker);
    await supabase.table('user_watchlists').update({ tickers }).eq('id', wl.id).eq('user_id', userId);
    return jsonResponse({ tickers }, 200, cors);
  }

  return jsonResponse({ tickers: [] }, 200, cors);
}
